File: android/app/src/main/java/com/openburnbar/data/hermes/relay/HermesDomainCoreAdapter.kt
package com.openburnbar.data.hermes.relay

import android.util.Log
import com.openburnbar.data.DomainCoreBuildProfile
import java.security.SecureRandom
import java.time.Instant
import uniffi.openburnbar_domain_ffi.HermesAadKind
import uniffi.openburnbar_domain_ffi.domainCoreAbiVersion
import uniffi.openburnbar_domain_ffi.domainCoreVersion
import uniffi.openburnbar_domain_ffi.hermesGatewayRelaySafetyCode
import uniffi.openburnbar_domain_ffi.hermesHkdfSha256
import uniffi.openburnbar_domain_ffi.hermesHmacSha256
import uniffi.openburnbar_domain_ffi.hermesHpkeV3Info
import uniffi.openburnbar_domain_ffi.hermesKeyWrapInfoV1
import uniffi.openburnbar_domain_ffi.hermesKeyWrapInfoV2
import uniffi.openburnbar_domain_ffi.hermesOpenBase64
import uniffi.openburnbar_domain_ffi.hermesOpenCombined
import uniffi.openburnbar_domain_ffi.hermesRatchetEnvelopeAad
import uniffi.openburnbar_domain_ffi.hermesRelayAad
import uniffi.openburnbar_domain_ffi.hermesSealBase64
import uniffi.openburnbar_domain_ffi.hermesSealCombined

internal enum class HermesDomainCoreMode {
    LEGACY,
    SHADOW,
    RUST,
    ;

    companion object {
        fun resolve(
            raw: String? = System.getProperty("openburnbar.domain_core.hermes.mode")
                ?: System.getenv("OPENBURNBAR_DOMAIN_CORE_HERMES_MODE"),
        ): HermesDomainCoreMode = entries.firstOrNull {
            it.name.equals(DomainCoreBuildProfile.mode("hermes", raw), ignoreCase = true)
        } ?: LEGACY
    }
}

internal data class HermesShadowComparison(
    val domain: String = "hermes",
    val slice: String,
    val consumer: String = "android",
    val operation: String,
    val coreVersion: String,
    val outcome: String,
    val mismatchCategory: String?,
    val legacyMicros: Long,
    val rustMicros: Long,
    val observedAt: Instant = Instant.now(),
)

internal object HermesDomainCoreAdapter {
    private const val REQUIRED_ABI_VERSION = 3u
    private const val ABI_MISMATCH_VERSION = "0.0.0-abi-mismatch"
    private const val NATIVE_UNAVAILABLE_VERSION = "0.0.0-native-unavailable"

    private enum class NativeAvailability(
        val diagnosticCategory: String,
        val evidenceVersion: String,
        val evidenceMismatchCategory: String,
    ) {
        READY("", "", ""),
        ABI_MISMATCH("abi_mismatch", ABI_MISMATCH_VERSION, "native_error"),
        NATIVE_UNAVAILABLE("native_unavailable", NATIVE_UNAVAILABLE_VERSION, "native_unavailable"),
    }

    private val secureRandom = SecureRandom()

    @Volatile
    internal var comparisonOverride: ((HermesShadowComparison) -> Unit)? = null

    @Volatile
    internal var modeOverride: HermesDomainCoreMode? = null

    @Volatile
    internal var abiVersionOverride: (() -> UInt)? = null

    @Volatile
    internal var coreVersionOverride: (() -> String)? = null

    fun aad(kind: HermesAadKind, arguments: List<String>, legacy: () -> ByteArray): ByteArray = selectBytes("aad", legacy) { hermesRelayAad(kind, arguments) }

    fun keyWrapInfoV1(aad: ByteArray, legacy: () -> ByteArray): ByteArray = selectBytes("key_wrap_info_v1", legacy) { hermesKeyWrapInfoV1(aad) }

    fun keyWrapInfoV2(aad: ByteArray, enc: ByteArray, recipient: ByteArray, sender: ByteArray, legacy: () -> ByteArray): ByteArray =
        selectBytes("key_wrap_info_v2", legacy) {
            hermesKeyWrapInfoV2(aad, enc, recipient, sender)
        }

    fun hpkeV3Info(aad: ByteArray, legacy: () -> ByteArray): ByteArray = selectBytes("hpke_v3_info", legacy) { hermesHpkeV3Info(aad) }

    fun hkdf(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int, legacy: () -> ByteArray): ByteArray = selectBytes("hkdf", legacy) {
        hermesHkdfSha256(ikm, salt, info, checkedHkdfLength(length))
    }

    fun seal(plaintext: ByteArray, key: ByteArray, aad: ByteArray, legacy: () -> String): String {
        val mode = resolvedMode()
        if (mode == HermesDomainCoreMode.LEGACY) return legacy()
        if (mode == HermesDomainCoreMode.SHADOW) {
            val legacyStarted = System.nanoTime()
            val old = legacy()
            val legacyMicros = elapsedMicros(legacyStarted)
            val availability = nativeAvailability()
            if (availability != NativeAvailability.READY) {
                collectUnavailable("seal", availability, legacyMicros)
                return old
            }
            val rustStarted = System.nanoTime()
            runCatching { hermesOpenBase64(old, key, aad) }.fold(
                onSuccess = { opened ->
                    val matches = opened.contentEquals(plaintext)
                    if (!matches) diagnostic("seal", "shadow_mismatch")
                    collect("seal", matches, if (matches) null else "result_mismatch", legacyMicros, elapsedMicros(rustStarted))
                },
                onFailure = {
                    diagnostic("seal", "native_error")
                    collect("seal", false, "native_error", legacyMicros, elapsedMicros(rustStarted))
                },
            )
            return old
        }
        val availability = nativeAvailability()
        if (availability != NativeAvailability.READY) return unavailable("seal", mode, availability, legacy)
        val nonce = ByteArray(12).also(secureRandom::nextBytes)
        return hermesSealBase64(plaintext, key, aad, nonce)
    }

    fun open(ciphertext: String, key: ByteArray, aad: ByteArray, legacy: () -> ByteArray): ByteArray =
        selectBytes("open", legacy) { hermesOpenBase64(ciphertext, key, aad) }

    fun safetyCode(agent: ByteArray, phone: ByteArray, legacy: () -> String): String =
        selectValue("safety_code", legacy, { hermesGatewayRelaySafetyCode(agent, phone) }, String::equals)

    fun hmac(key: ByteArray, data: ByteArray, operation: String, legacy: () -> ByteArray): ByteArray =
        selectBytes(operation, legacy) { hermesHmacSha256(key, data) }

    fun ratchetAad(header: HermesRatchetHeader, associatedData: ByteArray, legacy: () -> ByteArray): ByteArray = selectBytes("ratchet_aad", legacy) {
        hermesRatchetEnvelopeAad(
            associatedData,
            header.algorithm,
            header.sessionID,
            header.senderDeviceID,
            header.receiverDeviceID,
            header.ratchetPublicKeyBase64,
            header.version.toULong(),
            header.previousChainLength.toULong(),
            header.messageNumber.toULong(),
            header.epoch.toULong(),
        )
    }

    fun sealCombined(plaintext: ByteArray, key: ByteArray, aad: ByteArray, legacy: () -> ByteArray): ByteArray {
        val mode = resolvedMode()
        if (mode == HermesDomainCoreMode.LEGACY) return legacy()
        if (mode == HermesDomainCoreMode.SHADOW) {
            val legacyStarted = System.nanoTime()
            val old = legacy()
            val legacyMicros = elapsedMicros(legacyStarted)
            val availability = nativeAvailability()
            if (availability != NativeAvailability.READY) {
                collectUnavailable("ratchet_seal", availability, legacyMicros)
                return old
            }
            val rustStarted = System.nanoTime()
            runCatching { hermesOpenCombined(old, key, aad) }.fold(
                onSuccess = { opened ->
                    val matches = opened.contentEquals(plaintext)
                    if (!matches) diagnostic("ratchet_seal", "shadow_mismatch")
                    collect("ratchet_seal", matches, if (matches) null else "result_mismatch", legacyMicros, elapsedMicros(rustStarted))
                },
                onFailure = {
                    diagnostic("ratchet_seal", "native_error")
                    collect("ratchet_seal", false, "native_error", legacyMicros, elapsedMicros(rustStarted))
                },
            )
            return old
        }
        val availability = nativeAvailability()
        if (availability != NativeAvailability.READY) return unavailable("ratchet_seal", mode, availability, legacy)
        return hermesSealCombined(plaintext, key, aad, ByteArray(12).also(secureRandom::nextBytes))
    }

    fun openCombined(combined: ByteArray, key: ByteArray, aad: ByteArray, legacy: () -> ByteArray): ByteArray =
        selectBytes("ratchet_open", legacy) { hermesOpenCombined(combined, key, aad) }

    private fun selectBytes(operation: String, legacy: () -> ByteArray, rust: () -> ByteArray): ByteArray =
        selectValue(operation, legacy, rust, ByteArray::contentEquals)

    private fun <T> selectValue(operation: String, legacy: () -> T, rust: () -> T, equivalent: (T, T) -> Boolean): T {
        val mode = resolvedMode()
        if (mode == HermesDomainCoreMode.LEGACY) return legacy()
        if (mode == HermesDomainCoreMode.SHADOW) {
            val legacyStarted = System.nanoTime()
            val old = legacy()
            val legacyMicros = elapsedMicros(legacyStarted)
            val availability = nativeAvailability()
            if (availability != NativeAvailability.READY) {
                collectUnavailable(operation, availability, legacyMicros)
                return old
            }
            return selectValueWhenNativeAvailable(operation, mode, { old }, rust, equivalent)
        }
        val availability = nativeAvailability()
        if (availability != NativeAvailability.READY) return unavailable(operation, mode, availability, legacy)
        return selectValueWhenNativeAvailable(operation, mode, legacy, rust, equivalent)
    }

    internal fun <T> selectValueWhenNativeAvailable(
        operation: String,
        mode: HermesDomainCoreMode,
        legacy: () -> T,
        rust: () -> T,
        equivalent: (T, T) -> Boolean,
    ): T {
        if (mode == HermesDomainCoreMode.SHADOW) {
            val legacyStarted = System.nanoTime()
            val old = legacy()
            val legacyMicros = elapsedMicros(legacyStarted)
            val rustStarted = System.nanoTime()
            val value = runCatching(rust).getOrElse {
                diagnostic(operation, "native_error")
                collect(operation, false, "native_error", legacyMicros, elapsedMicros(rustStarted))
                return old
            }
            val matches = equivalent(old, value)
            if (!matches) diagnostic(operation, "shadow_mismatch")
            collect(operation, matches, if (matches) null else "result_mismatch", legacyMicros, elapsedMicros(rustStarted))
            return old
        }
        return rust()
    }

    internal fun checkedHkdfLength(length: Int): UInt {
        require(length in 1..(255 * 32)) { "Hermes HKDF output length is invalid" }
        return length.toUInt()
    }

    private fun resolvedMode(): HermesDomainCoreMode = modeOverride ?: HermesDomainCoreMode.resolve()

    internal fun resetTestOverrides() {
        comparisonOverride = null
        modeOverride = null
        abiVersionOverride = null
        coreVersionOverride = null
    }

    private fun nativeAvailability(): NativeAvailability {
        val abiVersion = runCatching { abiVersionOverride?.invoke() ?: domainCoreAbiVersion() }
            .getOrElse { return NativeAvailability.NATIVE_UNAVAILABLE }
        return if (abiVersion == REQUIRED_ABI_VERSION) NativeAvailability.READY else NativeAvailability.ABI_MISMATCH
    }

    private fun <T> unavailable(operation: String, mode: HermesDomainCoreMode, availability: NativeAvailability, legacy: () -> T): T {
        diagnostic(operation, availability.diagnosticCategory)
        if (mode == HermesDomainCoreMode.RUST) {
            throw IllegalStateException("Hermes Rust mode requires domain-core ABI v3")
        }
        return legacy()
    }

    private fun collectUnavailable(operation: String, availability: NativeAvailability, legacyMicros: Long) {
        check(availability != NativeAvailability.READY)
        diagnostic(operation, availability.diagnosticCategory)
        collect(
            operation = operation,
            matches = false,
            mismatchCategory = availability.evidenceMismatchCategory,
            legacyMicros = legacyMicros,
            rustMicros = 0,
            coreVersion = availability.evidenceVersion,
        )
    }

    private fun diagnostic(operation: String, outcome: String) {
        Log.w("OpenBurnBarDomainCore", "domain_core.hermes.$operation $outcome")
    }

    private fun elapsedMicros(startedNanos: Long): Long = ((System.nanoTime() - startedNanos) / 1_000)
        .coerceIn(0, 600_000_000)

    private fun collect(
        operation: String,
        matches: Boolean,
        mismatchCategory: String?,
        legacyMicros: Long,
        rustMicros: Long,
        coreVersion: String = safeCoreVersion(),
    ) {
        comparisonOverride?.invoke(
            HermesShadowComparison(
                slice = when {
                    operation == "aad" -> "aad"
                    operation.contains("hpke") -> "hpke-info"
                    operation.contains("ratchet") -> "ratchet"
                    else -> "payload-keywrap"
                },
                operation = operation,
                coreVersion = coreVersion,
                outcome = if (matches) "match" else "mismatch",
                mismatchCategory = mismatchCategory,
                legacyMicros = legacyMicros,
                rustMicros = rustMicros,
            ),
        )
    }

    private fun safeCoreVersion(): String = runCatching { coreVersionOverride?.invoke() ?: domainCoreVersion() }
        .getOrDefault(NATIVE_UNAVAILABLE_VERSION)
}

File: OpenBurnBarCore/Sources/OpenBurnBarKernel/SharedModels/HermesDomainCoreAdapter.swift
import Foundation
import OpenBurnBarDomainCoreRuntime

#if canImport(OpenBurnBarDomainCoreFFI)
import OpenBurnBarDomainCoreFFI
#endif

enum HermesDomainCoreMode: String, Sendable {
    case legacy
    case shadow
    case rust

    static func resolve(environment: [String: String] = ProcessInfo.processInfo.environment) -> Self {
        Self(rawValue: DomainCoreBuildProfileResolver.mode(for: .hermes, environment: environment).rawValue) ?? .legacy
    }
}

enum HermesDomainCoreAdapterError: Error {
    case nativeUnavailable
    case invalidInput
}

enum HermesDomainCoreAdapter {
    static var isNativeAvailable: Bool {
        #if canImport(OpenBurnBarDomainCoreFFI)
        OpenBurnBarDomainCoreFFI.domainCoreAbiVersion() == 3
        #else
        false
        #endif
    }

    static func aad(
        kind: HermesAadKindAdapter,
        arguments: [String],
        environment: [String: String] = ProcessInfo.processInfo.environment,
        legacy: () -> Data
    ) throws -> Data {
        try selectBytes(operation: "aad", environment: environment, legacy: { legacy() }, rust: {
            #if canImport(OpenBurnBarDomainCoreFFI)
            try OpenBurnBarDomainCoreFFI.hermesRelayAad(
                kind: kind.ffi,
                arguments: arguments
            )
            #else
            throw HermesDomainCoreAdapterError.nativeUnavailable
            #endif
        })
    }

    static func seal(
        plaintext: Data,
        key: Data,
        aad: Data,
        environment: [String: String] = ProcessInfo.processInfo.environment,
        legacy: () throws -> String
    ) throws -> String {
        let mode = HermesDomainCoreMode.resolve(environment: environment)
        return try DomainCoreShadowRuntime.selectVerifiedLegacy(
            domain: "hermes",
            slice: "payload-keywrap",
            operation: "seal",
            mode: DomainCoreBuildMode(rawValue: mode.rawValue) ?? .legacy,
            nativeStatus: nativeStatus,
            coreVersion: currentCoreVersion,
            legacy: legacy,
            rustAuthority: {
                #if canImport(OpenBurnBarDomainCoreFFI)
                try OpenBurnBarDomainCoreFFI.hermesSealBase64(
                    plaintext: plaintext,
                    key: key,
                    aad: aad,
                    nonce: PlatformCrypto.secureRandomBytes(count: 12)
                )
                #else
                throw HermesDomainCoreAdapterError.nativeUnavailable
                #endif
            },
            verifyLegacyWithRust: { ciphertext in
                #if canImport(OpenBurnBarDomainCoreFFI)
                try OpenBurnBarDomainCoreFFI.hermesOpenBase64(
                    ciphertext: ciphertext,
                    key: key,
                    aad: aad
                ) == plaintext
                #else
                throw HermesDomainCoreAdapterError.nativeUnavailable
                #endif
            },
            nativeUnavailableError: HermesDomainCoreAdapterError.nativeUnavailable,
            diagnostic: diagnostic
        )
    }

    static func open(
        ciphertext: String,
        key: Data,
        aad: Data,
        environment: [String: String] = ProcessInfo.processInfo.environment,
        legacy: () throws -> Data
    ) throws -> Data {
        try selectBytes(operation: "open", environment: environment, legacy: legacy) {
            #if canImport(OpenBurnBarDomainCoreFFI)
            try OpenBurnBarDomainCoreFFI.hermesOpenBase64(
                ciphertext: ciphertext,
                key: key,
                aad: aad
            )
            #else
            throw HermesDomainCoreAdapterError.nativeUnavailable
            #endif
        }
    }

    /// Combined AES-GCM seal for the relay key-wrap AEAD (raw bytes, no base64
    /// detour). Mirrors ``seal(plaintext:key:aad:environment:legacy:)`` but
    /// returns the combined ciphertext+tag `Data` the wrap path concatenates
    /// with the ephemeral public key. Shadow verifies the legacy output
    /// round-trips through Rust; in rust mode the Rust-produced combined bytes
    /// are returned and failures propagate without fallback.
    static func sealCombined(
        plaintext: Data,
        key: Data,
        aad: Data,
        environment: [String: String] = ProcessInfo.processInfo.environment,
        legacy: () throws -> Data
    ) throws -> Data {
        let mode = HermesDomainCoreMode.resolve(environment: environment)
        return try DomainCoreShadowRuntime.selectVerifiedLegacy(
            domain: "hermes",
            slice: "payload-keywrap",
            operation: "seal_combined",
            mode: DomainCoreBuildMode(rawValue: mode.rawValue) ?? .legacy,
            nativeStatus: nativeStatus,
            coreVersion: currentCoreVersion,
            legacy: legacy,
            rustAuthority: {
                #if canImport(OpenBurnBarDomainCoreFFI)
                try OpenBurnBarDomainCoreFFI.hermesSealCombined(
                    plaintext: plaintext,
                    key: key,
                    aad: aad,
                    nonce: PlatformCrypto.secureRandomBytes(count: 12)
                )
                #else
                throw HermesDomainCoreAdapterError.nativeUnavailable
                #endif
            },
            verifyLegacyWithRust: { combined in
                #if canImport(OpenBurnBarDomainCoreFFI)
                try OpenBurnBarDomainCoreFFI.hermesOpenCombined(
                    combined: combined,
                    key: key,
                    aad: aad
                ) == plaintext
                #else
                throw HermesDomainCoreAdapterError.nativeUnavailable
                #endif
            },
            nativeUnavailableError: HermesDomainCoreAdapterError.nativeUnavailable,
            diagnostic: diagnostic
        )
    }

    /// Combined AES-GCM open for the relay key-wrap AEAD (raw bytes). Mirrors
    /// ``open(ciphertext:key:aad:environment:legacy:)`` but operates on the
    /// combined ciphertext+tag `Data` the unwrap path splits off the envelope.
    /// Shadow compares Rust and legacy and returns legacy; in rust mode the
    /// Rust-produced plaintext is returned and failures propagate without
    /// fallback.
    static func openCombined(
        combined: Data,
        key: Data,
        aad: Data,
        environment: [String: String] = ProcessInfo.processInfo.environment,
        legacy: () throws -> Data
    ) throws -> Data {
        try selectBytes(operation: "open_combined", environment: environment, legacy: legacy) {
            #if canImport(OpenBurnBarDomainCoreFFI)
            try OpenBurnBarDomainCoreFFI.hermesOpenCombined(
                combined: combined,
                key: key,
                aad: aad
            )
            #else
            throw HermesDomainCoreAdapterError.nativeUnavailable
            #endif
        }
    }

    static func safetyCode(
        agent: Data,
        phone: Data,
        environment: [String: String] = ProcessInfo.processInfo.environment,
        legacy: () -> String
    ) throws -> String {
        let mode = HermesDomainCoreMode.resolve(environment: environment)
        return try DomainCoreShadowRuntime.select(
            domain: "hermes",
            slice: "payload-keywrap",
            operation: "safety_code",
            mode: DomainCoreBuildMode(rawValue: mode.rawValue) ?? .legacy,
            nativeStatus: nativeStatus,
            coreVersion: currentCoreVersion,
            legacy: legacy,
            rust: {
                #if canImport(OpenBurnBarDomainCoreFFI)
                try OpenBurnBarDomainCoreFFI.hermesGatewayRelaySafetyCode(
                    agentPublicKey: agent,
                    phonePublicKey: phone
                )
                #else
                throw HermesDomainCoreAdapterError.nativeUnavailable
                #endif
            },
            equivalent: ==,
            nativeUnavailableError: HermesDomainCoreAdapterError.nativeUnavailable,
            diagnostic: diagnostic,
            recordComparison: DomainCoreShadowComparisonCollector.record
        )
    }

    static func keyWrapInfoV1(
        aad: Data,
        environment: [String: String] = ProcessInfo.processInfo.environment,
        legacy: () throws -> Data
    ) throws -> Data {
        try selectBytes(operation: "key_wrap_info_v1", environment: environment, legacy: legacy) {
            #if canImport(OpenBurnBarDomainCoreFFI)
            try OpenBurnBarDomainCoreFFI.hermesKeyWrapInfoV1(aad: aad)
            #else
            throw HermesDomainCoreAdapterError.nativeUnavailable
            #endif
        }
    }

    static func keyWrapInfoV2(
        aad: Data,
        enc: Data,
        recipient: Data,
        sender: Data,
        environment: [String: String] = ProcessInfo.processInfo.environment,
        legacy: () throws -> Data
    ) throws -> Data {
        try selectBytes(operation: "key_wrap_info_v2", environment: environment, legacy: legacy) {
            #if canImport(OpenBurnBarDomainCoreFFI)
            try OpenBurnBarDomainCoreFFI.hermesKeyWrapInfoV2(
                aad: aad,
                enc: enc,
                recipientPublicKey: recipient,
                senderPublicKey: sender
            )
            #else
            throw HermesDomainCoreAdapterError.nativeUnavailable
            #endif
        }
    }

    static func hpkeV3Info(
        aad: Data,
        environment: [String: String] = ProcessInfo.processInfo.environment,
        legacy: () throws -> Data
    ) throws -> Data {
        try selectBytes(operation: "hpke_v3_info", environment: environment, legacy: legacy) {
            #if canImport(OpenBurnBarDomainCoreFFI)
            try OpenBurnBarDomainCoreFFI.hermesHpkeV3Info(aad: aad)
            #else
            throw HermesDomainCoreAdapterError.nativeUnavailable
            #endif
        }
    }

    static func hkdf(
        inputKeyMaterial: Data,
        salt: Data,
        info: Data,
        outputByteCount: Int,
        environment: [String: String] = ProcessInfo.processInfo.environment,
        legacy: () throws -> Data
    ) throws -> Data {
        return try selectBytes(operation: "hkdf", environment: environment, legacy: legacy) {
            #if canImport(OpenBurnBarDomainCoreFFI)
            try OpenBurnBarDomainCoreFFI.hermesHkdfSha256(
                inputKeyMaterial: inputKeyMaterial,
                salt: salt,
                info: info,
                outputByteCount: try checkedHkdfOutputByteCount(outputByteCount)
            )
            #else
            throw HermesDomainCoreAdapterError.nativeUnavailable
            #endif
        }
    }

    private static func selectBytes(
        operation: String,
        environment: [String: String],
        legacy: () throws -> Data,
        rust: () throws -> Data
    ) throws -> Data {
        let mode = HermesDomainCoreMode.resolve(environment: environment)
        return try selectBytes(
            operation: operation,
            mode: mode,
            nativeStatus: nativeStatus,
            coreVersion: currentCoreVersion,
            legacy: legacy,
            rust: rust,
            recordComparison: DomainCoreShadowComparisonCollector.record
        )
    }

    static func checkedHkdfOutputByteCount(_ value: Int) throws -> UInt32 {
        guard (1 ... 255 * 32).contains(value), let value = UInt32(exactly: value) else {
            throw HermesDomainCoreAdapterError.invalidInput
        }
        return value
    }

    static func selectBytes(
        operation: String,
        mode: HermesDomainCoreMode,
        nativeStatus: () -> DomainCoreRuntimeNativeStatus,
        coreVersion: () -> String,
        legacy: () throws -> Data,
        rust: () throws -> Data,
        recordComparison: (DomainCoreShadowComparison) -> Void
    ) throws -> Data {
        try DomainCoreShadowRuntime.select(
            domain: "hermes",
            slice: slice(for: operation),
            operation: operation,
            mode: DomainCoreBuildMode(rawValue: mode.rawValue) ?? .legacy,
            nativeStatus: nativeStatus,
            coreVersion: coreVersion,
            legacy: legacy,
            rust: rust,
            equivalent: ==,
            nativeUnavailableError: HermesDomainCoreAdapterError.nativeUnavailable,
            diagnostic: diagnostic,
            recordComparison: recordComparison
        )
    }

    private static func diagnostic(_ operation: String, _ outcome: String) {
        NSLog("domain_core.hermes.%@ %@", operation, outcome)
    }

    private static func slice(for operation: String) -> String {
        operation == "aad" ? "aad" : operation.contains("hpke") ? "hpke-info" : "payload-keywrap"
    }

    static func nativeStatus() -> DomainCoreRuntimeNativeStatus {
        #if canImport(OpenBurnBarDomainCoreFFI)
        if OpenBurnBarDomainCoreFFI.domainCoreAbiVersion() == 3 {
            return .available
        }
        return .unavailable(
            coreVersion: "0.0.0-abi-mismatch",
            mismatchCategory: "native_error"
        )
        #else
        return .unavailable(
            coreVersion: "0.0.0-native-unavailable",
            mismatchCategory: "native_unavailable"
        )
        #endif
    }

    static func currentCoreVersion() -> String {
        #if canImport(OpenBurnBarDomainCoreFFI)
        OpenBurnBarDomainCoreFFI.domainCoreVersion()
        #else
        "0.0.0-native-unavailable"
        #endif
    }
}

enum HermesAadKindAdapter {
    case request, key, authenticatedRequest, authenticatedKey, chunk
    case mediaSealKey, controlSealKey, gatewayEvent, gatewayEventKey
    case gatewayMessage, gatewayMessageKey, gatewayAttachmentKey
    case gatewayAttachmentManifest, gatewayAttachmentBody

    #if canImport(OpenBurnBarDomainCoreFFI)
    var ffi: OpenBurnBarDomainCoreFFI.HermesAadKind {
        switch self {
        case .request: .request
        case .key: .key
        case .authenticatedRequest: .authenticatedRequest
        case .authenticatedKey: .authenticatedKey
        case .chunk: .chunk
        case .mediaSealKey: .mediaSealKey
        case .controlSealKey: .controlSealKey
        case .gatewayEvent: .gatewayEvent
        case .gatewayEventKey: .gatewayEventKey
        case .gatewayMessage: .gatewayMessage
        case .gatewayMessageKey: .gatewayMessageKey
        case .gatewayAttachmentKey: .gatewayAttachmentKey
        case .gatewayAttachmentManifest: .gatewayAttachmentManifest
        case .gatewayAttachmentBody: .gatewayAttachmentBody
        }
    }
    #endif
}
